packages/firestore/src/util/bundle_builder_impl.ts

```diff
@@ -0,0 +1,290 @@
+    const originalQueries = originalDocument?.metadata.queries;
+
+    const readTime = docBundleData.readTime;
+    // Update with document built from `snap` because it is newer.
+    if (
+      !originalDocument ||
+      (!readTime && !originalDocument.metadata.readTime) ||
+      (readTime && originalDocument.metadata.readTime! < readTime)
+    ) {
+      this.documents.set(docBundleData.documentPath, {
+        document: this.toBundleDocument(docBundleData),
+        metadata: {
+          name: toName(this.serializer, docBundleData.documentKey),
+          readTime: !!readTime
+            ? toTimestamp(this.serializer, readTime)
+            : undefined,
+          exists: docBundleData.documentExists
+        }
+      });
+    }
+    // Update `queries` to include both original and `queryName`.
+    if (queryName) {
+      const newDocument = this.documents.get(docBundleData.documentPath)!;
+      newDocument.metadata.queries = originalQueries || [];
+      if (queryName) {
+        newDocument.metadata.queries!.push(queryName);
+      }
+    }
+    if (readTime && readTime > this.latestReadTime) {
+      this.latestReadTime = readTime;
+    }
+  }
+
+  /**
+   * Adds data from a QuerySnapshot to the bundle.
+   * @internal
+   * @param docBundleData A QuerySnapshotBundleData containing information from the
+   * QuerySnapshot. Note we cannot accept a QuerySnapshot directly due to a circular
+   * dependency error.
+   */
+  addBundleQuery(queryBundleData: QuerySnapshotBundleData): void {
+    const name = AutoId.newId();
+    if (this.namedQueries.has(name)) {
+      throw new Error(`Query name conflict: ${name} has already been added.`);
+    }
+    let latestReadTime = new Timestamp(0, 0);
+    for (const docBundleData of queryBundleData.docBundleDataArray) {
+      this.addBundleDocument(docBundleData, name);
+      if (docBundleData.readTime && docBundleData.readTime > latestReadTime) {
+        latestReadTime = docBundleData.readTime;
+      }
+    }
+    const queryTarget = toQueryTarget(
+      this.serializer,
+      queryToTarget(queryBundleData.query)
+    );
+    const bundledQuery = {
+      parent: queryBundleData.parent,
+      structuredQuery: queryTarget.queryTarget.structuredQuery
+    };
+    this.namedQueries.set(name, {
+      name,
+      bundledQuery,
+      readTime: toTimestamp(this.serializer, latestReadTime)
+    });
+  }
+
+  /**
+   * Convert data from a DocumentSnapshot into the serialized form within a bundle.
+   * @private
+   * @internal
+   * @param docBundleData a DocumentSnapshotBundleData containing the data required to
+   * serialize a document.
+   */
+  private toBundleDocument(
+    docBundleData: DocumentSnapshotBundleData
+  ): ProtoDocument {
+    // TODO handle documents that have mutations
+    debugAssert(
+      !docBundleData.documentData.hasLocalMutations,
+      "Can't serialize documents with mutations."
+    );
```

Comment:
We will have to consider the impact of this assert on serializing documents with local mutations. This could be a breaking change if someone is already performing `JSON.stringify(mutatedDoc)`. This is a use case that should not matter for SSR, but will matter because we are using `toJSON` for serialization.

Reply:
I guess there's also a case where someone writes or modifies a document on the server side, and then runs a query to get that document (with modifications) before the server gets a write ack from the backend. This means the documentsnapshot would be with mutations...

I'm not even sure how SSR should behave in this case.

Reply:
This is a tricky situation for SSR. We will have to be clear that the server side code must use `getDocumentsFromServer` to ensure that we don't get a document with local mutation.

Also, I'm beginning to wonder if we should only override toJSON for these snapshot types when running server side. This helps avoid dealing with edge cases of serialization in the client side. Thoughts?

Reply:
We discussed this offline, and resolved to throw in these situations. I've updated the branch accordingly. PTAL.